Fix Pull Up/Down intermittently not working

krwq opened the pull request:
Fixes #943
Contributes to #941

Did couple of changes before all failures were gone. Out of 50 CI runs 1 failed and issue was unrelated (the other intermittent failure from #941).

Tested only on Release build.

When testing number of iterations in the test was increased to 10k

Changes made:
- added zeroing clock (this fixed most of the problems)
- operating directly on pointer
- despite spec order zeroing clock happens first since that seem to have causing intermittent issues as well
- SpinWait was replaced with regular loop and optimizations in that method were disabled - we can optimize that in the future but I wasn't able to get no intermittent failures without this also was paranoid that pointer operations might get optimized in some way to use local in some hidden away - not sure if that made any difference

cc: @pgrawehr

krwq commented:
@pgrawehr likely similar change has to be done on PI4 but I don't have a setup at the moment where I could test it

pgrawehr commented:
Thanks for investigating this. I'll play around with my Pi4 tonight, to see what fixes the issue there. The code does not currently contain any wait magic there, as setting the pulls is now similarly easy as setting the value. However, the official CPU register documentation for the Pi4 has still not been released, so there might be some wait required that's just undocumented.

joperezr commented on src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.Linux.cs:
```diff
@@ -258,28 +259,28 @@ private void SetInputPullMode(int pinNumber, PinMode mode)
+            *gppudclkPointer &= ~pinBit;
+            *gppudPointer &= ~0b11U;
+
+            for (int i = 0; i < 150; i++)
```
Can you please update the comment above the process to point out that we had to go against spec because that was the way to make it work? That way if we ever have to change this we will take that into consideration.

krwq replied:
actually I'm not sure this wait is needed, I added this when testing - will double check...

krwq replied:
seems the delay might be needed - I'm getting 2/50 failures without it. Will double check my original results to make sure I wasn't just lucky with no failures

krwq replied:
100/100 passed with this timeout, 2/50 failed without it

joperezr commented on src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.Linux.cs:
```diff
@@ -226,6 +226,7 @@ protected internal override void SetPinMode(int pinNumber, PinMode mode)
         /// </summary>
         /// <param name="pinNumber">The pin number in the driver's logical numbering scheme.</param>
         /// <param name="mode">The mode of a pin to set the resistor pull up/down mode.</param>
+        [MethodImpl(MethodImplOptions.NoOptimization)]
```
Can you put a comment on top of this option to mention why do we require noOptimization here?

krwq replied:
I've put it inside since the comment looked weird under the docs

joperezr commented:
Great find @krwq I'm glad you were able to fix this so fast :)

pgrawehr commented:
Something is weird. I'm (correctly) getting StyleCop errors when building using VS, and also when I'm building `System.Devices.Gpio.sln` using dotnet build directly, but not when executing the batch.

pgrawehr commented:
There's a problem with the build configuration settings that make Stylecop not run by default. It seems that to fix that, I need to find out where and how `System.Device.Gpio.notsupported.cs` is generated (because that autogenerated file is not StyleCop compliant and fails to declare itself as autogenerated).

pgrawehr commented:
@krwq: Addressing the StyleCop stuff in #947, but please build once in VS to fix the issues you are about to introduce.
Note: Empty statements should be written as
```
for (...)
{
}
```

pgrawehr commented on src/System.Device.Gpio/System/Device/Gpio/Drivers/RaspberryPi3Driver.Linux.cs:
```diff
@@ -276,6 +282,9 @@ private void SetInputPullMode(int pinNumber, PinMode mode)
 
+            // Spec calls to reset clock after the control signal
+            // Since context switch between those two instructions can potentially
+            // change pull up/down value we reset the clock first.
```
Talking about context switches... What happens if two threads try to simultaneously change two (distinct) pins at the same time? Do we need a global lock around this method?

krwq commented:
addressed all comments so will merge after CI is green. We should eventually investigate if we can simplify this logic but there is no pressure since this is not a hot path
